### PR TITLE
Wenqi/cypress setup

### PR DESCRIPTION
**Pull request by hwenchi**

### CODE REVIEW

**Comment by matthewberry** [comment]
Hi Wenqi, looks great!

Now that you've worked with cypress a bit, does it seem like a good solution for testing? Did you see any cypress features we're not using yet but should be thinking about using in the future? Any answer is fine, and we can always investigate more later, but I'm curious to hear any impressions you have at this point.

**Review by matthewberry — APPROVED** [review]

**Comment by matthewberry** [comment]
Approved. Feel free to merge whenever you're ready, Wenqi.

File: package.json [hunk 1]
```diff
@@ -7,6 +7,8 @@
     "build": "ng build",
     "watch": "ng build --watch --configuration development",
     "test": "ng test",
+    "test:e2e": "run-p start cypress",
```
Comment (matthewberry) [inline package.json:10]: Could you please add a brief note to the project readme about these commands?

**Review by matthewberry — APPROVED** [review]

**Comment by hwenchi** [comment]
> Hi Wenqi, looks great!
> 
> Now that you've worked with cypress a bit, does it seem like a good solution for testing? Did you see any cypress features we're not using yet but should be thinking about using in the future? Any answer is fine, and we can always investigate more later, but I'm curious to hear any impressions you have at this point.

I think Cypress is pretty nice. There aren't any test cases involving data / requests yet, and it looks like [intercept](https://docs.cypress.io/api/commands/intercept) and [fixture](https://docs.cypress.io/api/commands/fixture#Syntax) could be useful for those. Also, I saw that Cypress supports component testing, but Angular also comes with its own component testing package. We might need to decide which one to use.

File: package.json (hunk 1 above)
Comment (hwenchi) [reply package.json:10]: Sure

Comment (matthewberry) [reply package.json:10]: Thanks

**Comment by matthewberry** [comment]
Thanks for your thoughts on Cypress. I've barely scratched the surface but have seen a couple things that might come in handy:
- Cypress Studio (https://docs.cypress.io/guides/references/cypress-studio#Adding-a-New-Test) offers a way to record user interactions into tests, which could make it easier for, e.g., Charles to help us create cases.
- There are plugins for visual regression testing (e.g., https://www.npmjs.com/package/cypress-visual-regression, https://www.npmjs.com/package/cypress-plugin-snapshots), which might make it easier to validate the state of something as complex as a multi-tree visualization, especially if we end up rendering to `<canvas>` or something like that.

Do you have any experience with other tools along those lines?

**Comment by hwenchi** [comment]
> Thanks for your thoughts on Cypress. I've barely scratched the surface but have seen a couple things that might come in handy:
> 
>     * Cypress Studio (https://docs.cypress.io/guides/references/cypress-studio#Adding-a-New-Test) offers a way to record user interactions into tests, which could make it easier for, e.g., Charles to help us create cases.
> 
>     * There are plugins for visual regression testing (e.g., https://www.npmjs.com/package/cypress-visual-regression, https://www.npmjs.com/package/cypress-plugin-snapshots), which might make it easier to validate the state of something as complex as a multi-tree visualization, especially if we end up rendering to `<canvas>` or something like that.
> 
> 
> Do you have any experience with other tools along those lines?

- I remember [Playwright Codegen](https://playwright.dev/docs/codegen) is used at my previous company. They look quite similar, although I believe Playwright operates at a lower (Chrome DevTools Protocol) level and not in browser like Cypress. 
- I don't have any experience with visual regression testing yet. (In a graphics course I'm taking, we use [ImageMagick](https://imagemagick.org/script/compare.php) to do fuzzy image diffing for the purpose of testing/grading. I feel the idea is kind of similar.)

**Pull request merged**